rules/CodeQuality/Rector/ClassConstFetch/ConvertStaticPrivateConstantToSelfRector.php

```diff
@@ -0,0 +1,118 @@
+<?php
+
+declare(strict_types=1);
+
+namespace Rector\CodeQuality\Rector\ClassConstFetch;
+
+use Exception;
+use PhpParser\Node;
+use PhpParser\Node\Expr\ClassConstFetch;
+use PhpParser\Node\Expr\Error;
+use PhpParser\Node\Name;
+use PHPStan\Analyser\Scope;
+use PHPStan\Reflection\ClassReflection;
+use Rector\Core\Rector\AbstractRector;
+use Rector\Core\Reflection\ReflectionResolver;
+use Rector\NodeTypeResolver\Node\AttributeKey;
+use Symplify\RuleDocGenerator\ValueObject\CodeSample\CodeSample;
+use Symplify\RuleDocGenerator\ValueObject\RuleDefinition;
+
+/**
+ * @see \Rector\Tests\CodeQuality\Rector\ClassConstFetch\ConvertStaticPrivateConstantToSelfRector\ConvertStaticPrivateConstantToSelfRectorTest
+ * @see https://3v4l.org/8Y0ba
+ * @see https://phpstan.org/r/11d4c850-1a40-4fae-b665-291f96104d11
+ */
+final class ConvertStaticPrivateConstantToSelfRector extends AbstractRector
+{
+    public function __construct(
+        private readonly ReflectionResolver $reflectionResolver
+    ) {
+    }
+
+    public function getRuleDefinition(): RuleDefinition
+    {
+        return new RuleDefinition(
+            'Replaces static::* access to private constants with self::* on final classes',
+            [
+                new CodeSample(
+                    <<<'CODE_SAMPLE'
+final class Foo {
+    private const BAR = 'bar';
+    public function run()
+    {
+        $bar = static::BAR;
+    }
+}
+CODE_SAMPLE
+                    ,
+                    <<<'CODE_SAMPLE'
+final class Foo {
+    private const BAR = 'bar';
+    public function run()
+    {
+        $bar = self::BAR;
+    }
+}
+CODE_SAMPLE
+                    ,
+                ),
+            ],
+        );
+    }
+
+    public function getNodeTypes(): array
+    {
+        return [ClassConstFetch::class];
+    }
+
+    /**
+     * @param \PhpParser\Node\Expr\ClassConstFetch $node
+     */
+    public function refactor(Node $node): ?ClassConstFetch
+    {
+        if (! $this->isUsingStatic($node)) {
+            return null;
+        }
+
+        if (! $this->isPrivateConstant($node)) {
+            return null;
+        }
+
+        return new ClassConstFetch(new Name('self'), $node->name);
+    }
+
+    private function isUsingStatic(ClassConstFetch $node): bool
+    {
+        if (! $node->class instanceof Name) {
+            return false;
+        }
+
+        return $node->class->getFirst() === 'static';
+    }
+
+    private function isPrivateConstant(ClassConstFetch $node): bool
+    {
+        $scope = $node->getAttribute(AttributeKey::SCOPE);
+        if (! $scope instanceof Scope) {
+            return false;
+        }
+        $classReflection = $this->reflectionResolver->resolveClassReflection($node);
+        if (! $classReflection instanceof ClassReflection) {
+            return false;
+        }
+        if (! $classReflection->isClass() || ! $classReflection->isFinal()) {
```

Comment:
`$classReflection->isFinal()` should not be used for this case as it marking `@final` docblock as final class, use `Class_` node ->isFinal() instead that verify native `final` modifier.

Reply:
Done